mlflow/pyfunc/model.py

```diff
@@ -0,0 +1,367 @@
+
+
+class PythonModel(object):
+    """
+    Represents a generic Python model that leverages a collection of *artifacts* and *parameters*
+    (Python objects) to evaluate inputs and produce API-compatible outputs. By subclassing
+    :class:`~PythonModel`, users can create customized MLflow models with the "python_function"
+    ("pyfunc") flavor, leveraging custom inference logic and dependencies.
+    """
+    __metaclass__ = ABCMeta
+
+    def __init__(self, context):
+        """
+        :param context: A :class:`~PythonModelContext` instance containing *artifacts* and
+                        *parameters* that the model can use to perform inference.
+        """
+        self.context = context
+
+    @abstractmethod
+    def predict(self, model_input):
+        """
+        Evaluates a pyfunc-compatible input and produces a pyfunc-compatible output.
+        For more information about the pyfunc input/output API, see `Inference API`_.
+
+        :param model_input: A pyfunc-compatible input for the model to evaluate.
+        """
+        pass
+
+
+class PythonModelContext(object):
+    """
+    A collection of *artifacts* and *parameters* (Python objects) that a :class:`~PythonModel` can
+    use when performing inference. :class:`~PythonModelContext` objects are created *implicitly* by
+    the :func:`save_model() <mlflow.pyfunc.save_model>` and
+    :func:`log_model() <mlflow.pyfunc.log_model>` persistence methods, using the contents specified
+    by the ``artifacts`` and ``parameters`` arguments of these methods.
+    """
+
+    def __init__(self, artifacts, parameters, directory_managers=None):
+        """
+        :param artifacts: A dictionary of ``<name, artifact_path>`` entries, where ``artifact_path``
+                          is an absolute filesystem path to a given artifact.
+        :param parameters: A dictionary of ``<name, python object>`` entries.
+        :param directories: A list of objects managing the lifecycle of directories that the
+                            contain the specified artifacts. These objects will be stored
+                            as a class attribute, ensuring that their associated directories
+                            exist for as long as the :class:`~PythonModelContext` is in scope.
+        """
+        self._artifacts = artifacts
+        self._parameters = parameters
+        self._directory_managers = directory_managers
+
+    @property
+    def artifacts(self):
+        """
+        :return: A dictionary containing ``<name, artifact_path>`` entries, where ``artifact_path``
+                 is an absolute filesystem path to the artifact.
+        """
+        return self._artifacts
+
+    @property
+    def parameters(self):
+        """
+        :return: A dictionary containing ``<name, python object>`` entries.
+        """
+        return self._parameters
+
+
+def _save_model_with_class_artifacts_params(path, model_class, artifacts=None, parameters=None,
```

Comment:
👍 on how you broke this down into class vs loader based and forking it into individual flows! Awesome!

Reply:
Thanks Mani!